source/delegate/contracts/DelegateV2.sol

```diff
@@ -0,0 +1,543 @@
+
+    // If no delegate owner is provided, the deploying address is the owner.
+    if (delegateContractOwner != address(0)) {
+      transferOwnership(delegateContractOwner);
+    }
+
+    // If no trade wallet is provided, the owner's wallet is the trade wallet.
+    if (delegateTradeWallet != address(0)) {
+      tradeWallet = delegateTradeWallet;
+    } else {
+      tradeWallet = owner();
+    }
+
+    // Ensure that the indexer can pull funds from delegate account.
+    require(
+      IERC20(indexer.stakingToken()).approve(address(indexer), MAX_INT),
+      "STAKING_APPROVAL_FAILED"
+    );
+  }
+
+  function createRule(
+    address senderToken,
+    address signerToken,
+    uint256 senderAmount,
+    uint256 signerAmount
+  ) external onlyOwner {
+    _createRule(senderToken, signerToken, senderAmount, signerAmount);
+  }
+
+  function deleteRule(uint256 ruleID) external onlyOwner {
+    require(rules[ruleID].senderAmount != 0, "RULE_NOT_ACTIVE");
+    _deleteRule(ruleID);
+  }
+
+  function createRuleAndSetIntent(
+    address senderToken,
+    address signerToken,
+    uint256 senderAmount,
+    uint256 signerAmount,
+    uint256 newStakeAmount
+  ) external onlyOwner {
+    // create rule and
+    _createRule(senderToken, signerToken, senderAmount, signerAmount);
+
+    // set intent
+    _setIntent(senderToken, signerToken, newStakeAmount);
+  }
+
+  function setIntent(
+    address senderToken,
+    address signerToken,
+    uint256 newStakeAmount
+  ) external onlyOwner {
+    // check whether the delegate has a rule for that market
+    require(firstRuleID[senderToken][signerToken] != NO_RULE, "NO_RULE_EXISTS");
+
+    _setIntent(senderToken, signerToken, newStakeAmount);
+  }
+
+  function deleteRuleAndUnsetIntent(uint256 ruleID) external onlyOwner {
+    require(rules[ruleID].senderAmount != 0, "RULE_NOT_ACTIVE");
+
+    address signerToken = rules[ruleID].signerToken;
+    address senderToken = rules[ruleID].senderToken;
+
+    _deleteRule(ruleID);
+
+    _unsetIntent(senderToken, signerToken);
+  }
+
+  function unsetIntent(address senderToken, address signerToken)
+    external
+    onlyOwner
+  {
+    _unsetIntent(senderToken, signerToken);
+  }
+
+  function provideOrder(Types.Order calldata order) external {
+    uint256 ruleID = firstRuleID[order.sender.token][order.signer.token];
+
+    require(order.signature.v != 0, "SIGNATURE_MUST_BE_SENT");
+
+    // Ensure the order is for the trade wallet.
+    require(order.sender.wallet == tradeWallet, "SENDER_WALLET_INVALID");
+
+    // Ensure the tokens are valid ERC20 tokens.
+    require(
+      order.signer.kind == ERC20_INTERFACE_ID,
+      "SIGNER_KIND_MUST_BE_ERC20"
+    );
+
+    require(
+      order.sender.kind == ERC20_INTERFACE_ID,
+      "SENDER_KIND_MUST_BE_ERC20"
+    );
+
+    // Ensure that a rule exists.
+    require(ruleID != NO_RULE, "TOKEN_PAIR_INACTIVE");
+
+    uint256 signerAmount = order.signer.amount;
+    uint256 senderAmount = order.sender.amount;
```

Comment:
Suggest `remainingSignerAmount` and `remainingSenderAmount` as below

Reply:
Done